src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,420 @@
+                                     const QString &name,
+                                     const QString &value)
+{
+    if (!value.isEmpty()) {
+        stream.writeAttribute(name, value);
+    }
+}
+
+void XmlMapStorage::log(const QString &msg)
+{
+    emit sig_log("XmlMapStorage", msg);
+}
+
+#define DECL(X, ...) #X,
+static const char *const alignNames[] = {
+    X_FOREACH_RoomAlignEnum(DECL) //
+};
+static const char *const doorFlagNames[] = {
+    X_FOREACH_DOOR_FLAG(DECL) //
+};
+static const char *const exitFlagNames[] = {
+    X_FOREACH_EXIT_FLAG(DECL) //
+};
+static const char *const lightNames[] = {
+    X_FOREACH_RoomLightEnum(DECL) //
+};
+static const char *const loadFlagNames[] = {
+    X_FOREACH_ROOM_LOAD_FLAG(DECL) //
+};
+static const char *const markClassNames[] = {
+    X_FOREACH_INFOMARK_CLASS(DECL) //
+};
+static const char *const markTypeNames[] = {
+    X_FOREACH_INFOMARK_TYPE(DECL) //
+};
+static const char *const mobFlagNames[] = {
+    X_FOREACH_ROOM_MOB_FLAG(DECL) //
+};
+static const char *const portableNames[] = {
+    X_FOREACH_RoomPortableEnum(DECL) //
+};
+static const char *const ridableNames[] = {
+    X_FOREACH_RoomRidableEnum(DECL) //
+};
+static const char *const sundeathNames[] = {
+    X_FOREACH_RoomSundeathEnum(DECL) //
+};
+static const char *const terrainNames[] = {
+    X_FOREACH_RoomTerrainEnum(DECL) //
+};
+#undef DECL
+
+void XmlMapStorage::saveDoorFlags(QXmlStreamWriter &stream, DoorFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    QString list;
+    const char *separator = "";
+    for (uint x = 0; x < sizeof(doorFlagNames) / sizeof(doorFlagNames[0]); x++) {
+        if (fl.contains(DoorFlagEnum(x))) {
+            list += separator;
+            separator = ",";
+            list += doorFlagNames[x];
+        }
+    }
+    saveXmlAttribute(stream, "doorflags", list);
+}
+
+void XmlMapStorage::saveExitFlags(QXmlStreamWriter &stream, ExitFlags fl)
+{
+    fl.remove(ExitFlagEnum::EXIT); // always set, do not save it
+    if (fl.isEmpty()) {
+        return;
+    }
+    QString list;
+    const char *separator = "";
+    for (uint x = 0; x < sizeof(exitFlagNames) / sizeof(exitFlagNames[0]); x++) {
+        if (fl.contains(ExitFlagEnum(x))) {
+            list += separator;
+            separator = ",";
+            list += exitFlagNames[x];
+        }
+    }
+    saveXmlAttribute(stream, "flags", list);
+}
+
+void XmlMapStorage::saveRoomLoadFlags(QXmlStreamWriter &stream, RoomLoadFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    stream.writeStartElement("flags");
+    QString separator;
+    for (uint x = 0; x < sizeof(loadFlagNames) / sizeof(loadFlagNames[0]); x++) {
+        if (fl.contains(RoomLoadFlagEnum(x))) {
+            stream.writeCharacters(separator);
+            if (separator.isEmpty()) {
+                separator = ",";
+            }
+            stream.writeCharacters(loadFlagNames[x]);
```

Comment:
Why not have each flag be a separate XML element?

Reply:
That's certainly possible. It could possibly generate lots of lines though.
But on the other hand it would be much easier to validate with a DTD or XML Schema (see next item)

Reply:
Done. RoomLoadFlags, DoorFlags and ExitFlags are now saved as repeated XML elements instead of a comma-separated list inside the value of an XML attribute